app/post_persistent_notification.go

```diff
@@ -0,0 +1,380 @@
+			}
+			channelGroupMap[c.Id] = make(map[string]*model.Group, len(groups))
+			for k, v := range groups {
+				channelGroupMap[c.Id][k] = v
+			}
+			props, err := a.Srv().Store().Channel().GetAllChannelMembersNotifyPropsForChannel(c.Id, true)
+			if err != nil {
+				return errors.Wrap(err, "failed to get channel notify props")
+			}
+			channelNotifyProps[c.Id] = props
+		}
+
+		profileMap, err := a.Srv().Store().User().GetAllProfilesInChannel(context.Background(), c.Id, true)
+		if err != nil {
+			return errors.Wrapf(err, "failed to get profiles for channel %s", c.Id)
+		}
+
+		channelKeywords[c.Id] = make(map[string][]string, len(profileMap))
+		validProfileMap := make(map[string]*model.User, len(profileMap))
+		for k, v := range profileMap {
+			if v.IsBot {
+				continue
+			}
+			validProfileMap[k] = v
+			channelKeywords[c.Id]["@"+v.Username] = []string{k}
+		}
+		channelProfileMap[c.Id] = validProfileMap
+	}
+
+	for _, post := range posts {
+		channel := channelsMap[post.ChannelId]
+		team := teamsMap[channel.TeamId]
+		if channel.IsGroupOrDirect() {
+			team = &model.Team{}
+		}
+		profileMap := channelProfileMap[channel.Id]
+
+		mentions := &ExplicitMentions{}
+		if channel.Type == model.ChannelTypeDirect {
+			otherUserId := channel.GetOtherUserIdForDM(post.UserId)
+			if _, ok := profileMap[otherUserId]; ok {
+				mentions.addMention(otherUserId, DMMention)
+			}
+		} else {
+			keywords := channelKeywords[channel.Id]
+			mentions = getExplicitMentions(post, keywords, channelGroupMap[channel.Id])
+			for _, group := range mentions.GroupMentions {
+				_, err := a.insertGroupMentions(group, channel, profileMap, mentions)
+				if err != nil {
+					return errors.Wrapf(err, "failed to include mentions from group - %s for channel - %s", group.Id, channel.Id)
+				}
+			}
+		}
+
+		if err := fn(post, channel, team, mentions, profileMap, channelNotifyProps); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (a *App) sendPersistentNotifications(post *model.Post, channel *model.Channel, team *model.Team, mentions *ExplicitMentions, profileMap model.UserMap, channelNotifyProps map[string]map[string]model.StringMap) error {
+	mentionedUsersList := make(model.StringArray, 0, len(mentions.Mentions))
+	for id := range mentions.Mentions {
+		// Don't send notification to post owner
+		if id != post.UserId {
+			mentionedUsersList = append(mentionedUsersList, id)
+		}
+	}
+
+	sender := profileMap[post.UserId]
+	notification := &PostNotification{
+		Post:       post,
+		Channel:    channel,
+		ProfileMap: profileMap,
+		Sender:     sender,
+	}
+
+	// Check for channel-wide mentions in channels that have too many members for those to work
+	if int64(len(mentionedUsersList)) > *a.Config().TeamSettings.MaxNotificationsPerChannel {
+		return errors.Errorf("mentioned users: %d are more than allowed users: %d", len(mentionedUsersList), *a.Config().TeamSettings.MaxNotificationsPerChannel)
+	}
```

Comment:
I don't understand this comment: what does the size of the channel have to do with the restrictions?

Reply:
The comment may not be needed here, I'll remove it. The comment was for `@all` and similar channel-wide tags.

As for push-notifications, there is this max. limit for any channel - `MaxNotificationsPerChannel`.